system/lib/wasmfs/proxied_file_backend.cpp

```diff
@@ -0,0 +1,101 @@
+// Copyright 2021 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+
+// This file defines the Proxied File and Proxied Backend of the new file
+// system. Current Status: Work in Progress. See
+// https://github.com/emscripten-core/emscripten/issues/15041.
+
+#include "backend.h"
+#include "file.h"
+#include "thread_utils.h"
+#include "wasmfs.h"
+
+namespace wasmfs {
+
+// This class represents a file that is resident on a background pthread. All
+// file operations are proxied to this dedicated thread.
+class ProxiedFile : public DataFile {
+
+  emscripten::SyncToAsync& proxy;
+  std::shared_ptr<DataFile> baseFile;
+
+  // Read and write operations are forwarded to the file residing on the
+  // dedicated thread.
+  __wasi_errno_t write(const uint8_t* buf, size_t len, off_t offset) override {
+    __wasi_errno_t result;
+    proxy.invoke([&](emscripten::SyncToAsync::Callback resume) {
+      result = baseFile->locked().write(buf, len, offset);
+      (*resume)();
+    });
+    return result;
+  }
+
+  __wasi_errno_t read(uint8_t* buf, size_t len, off_t offset) override {
+    __wasi_errno_t result;
+    proxy.invoke([&](emscripten::SyncToAsync::Callback resume) {
+      result = baseFile->locked().read(buf, len, offset);
+      (*resume)();
+    });
+    return result;
+  }
+
+  // Querying the size of the Proxied File returns the size of the underlying
+  // file on the dedicated thread.
+  size_t getSize() override {
+    size_t result;
+    proxy.invoke([&](emscripten::SyncToAsync::Callback resume) {
+      result = baseFile->locked().getSize();
+      (*resume)();
+    });
+    return result;
+  }
+
+public:
+  // A file with the chosen destination backend is created on the Proxied File's
+  // dedicated thread.
+  ProxiedFile(mode_t mode,
+              backend_t backend,
+              std::shared_ptr<DataFile> file,
+              emscripten::SyncToAsync& proxy)
+    : DataFile(mode, backend), proxy(proxy), baseFile(file) {}
+
+  // The destructor must proxy to the pthread and release the file stored on
+  // that pthread.
+  ~ProxiedFile() {
+    proxy.invoke([&](emscripten::SyncToAsync::Callback resume) {
+      baseFile = nullptr;
+      (*resume)();
+    });
+  }
```

Comment:
I'm not sure what this destructor is doing. `baseFile` is just a property on this object, so it will be cleared on exit - is there a reason we need to do it on a particular thread? (I don't see one.)

Reply:
Do we need to do anything to free `JSFile` that is using the this proxying backend?

Reply:
Someone needs to do that, yeah. But if this class is responsible then I'd expect to see it both create and destroy the file on the same thread, and I don't think I see that? It just proxies the destructor.

Reply:
We discussed this and I will add some comments about proxying to a thread to free the underlying resource. Ex. `JSFile` will need to specifically free its JS array on that proxied thread.

Reply:
I agree that it would be nice to have this class responsible for all the proxying, including for creation and deletion. If we went back to supplying the underlying backend to the `ProxiedFile` constructor, the constructor could proxy the creation of the underlying `file` to the target thread as well.

Reply:
Sounds good, updated this to move the proxying back to file constructor. Updated comments.